zipkin-storage/elasticsearch/src/main/java/zipkin/storage/elasticsearch/ElasticsearchSpanConsumer.java

```diff
@@ -73,14 +75,39 @@
 
   private IndexRequestBuilder createSpanIndexRequest(Span input) {
     Span span = ApplyTimestampAndDuration.apply(input);
-    long indexTimestampMillis;
+    long timestampMillis;
+    final byte[] spanBytes;
     if (span.timestamp != null) {
-      indexTimestampMillis = TimeUnit.MICROSECONDS.toMillis(span.timestamp);
+      timestampMillis = TimeUnit.MICROSECONDS.toMillis(span.timestamp);
+      spanBytes = prefixWithTimestampMillis(Codec.JSON.writeSpan(span), timestampMillis);
     } else {
-      indexTimestampMillis = System.currentTimeMillis();
+      timestampMillis = System.currentTimeMillis();
+      spanBytes = Codec.JSON.writeSpan(span);
     }
-    String spanIndex = indexNameFormatter.indexNameForTimestamp(indexTimestampMillis);
+    String spanIndex = indexNameFormatter.indexNameForTimestamp(timestampMillis);
     return client.prepareIndex(spanIndex, ElasticsearchConstants.SPAN)
-        .setSource(Codec.JSON.writeSpan(span));
+        .setSource(spanBytes);
+  }
+
+  /**
+   * In order to allow systems like Kibana to search by timestamp, we add a field "timestamp_millis"
+   * when storing. The cheapest way to do this without changing the codec is prefixing it to the
+   * json. For example. {"traceId":"... becomes {"timestamp_millis":12345,"traceId":"...
+   */
+  @VisibleForTesting
+  static byte[] prefixWithTimestampMillis(byte[] input, long timestampMillis) {
```

Comment:
Heh - modifying the JSON was supposed to be the annoying part of this, but I didn't think of this reasonable(-ish) solution :)